3p-ecosystem/src/main/java/com/google/homesampleapp/screens/thread/ThreadFragment.kt

```diff
@@ -130,66 +200,703 @@ class ThreadFragment : Fragment() {
 
   /** Setup Observers */
   private fun setupObservers() {
-    /** Registers for activity result from Google Play Services */
+    // Registers for activity result from Google Play Services.
+    // Note that Fragments must call registerForActivityResult() before they are created
     threadClientLauncher =
-        registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
-          if (result.resultCode == RESULT_OK) {
-            val threadNetworkCredentials =
-                ThreadNetworkCredentials.fromIntentSenderResultData(result.data!!)
-            viewModel.threadPreferredCredentialsOperationalDataset.postValue(
-                threadNetworkCredentials)
-          } else {
-            val error = "User denied request."
-            Timber.d(error)
-            updateThreadInfo(null, "")
+      registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
+        if (result.resultCode == RESULT_OK) {
+          val threadNetworkCredentials =
+            ThreadNetworkCredentials.fromIntentSenderResultData(result.data!!)
+          viewModel.setThreadCredentialsInfo(null, threadNetworkCredentials)
+        } else {
+          val error = "User denied request."
+          Timber.d(error)
+          viewModel.setThreadCredentialsInfo(null, null)
+        }
+      }
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Top level Composables
+
+  @Composable
+  private fun ThreadRoute() {
+    // UI Logic implemented in ThreadNetworkUiState requires Activity.
+    val activity = LocalContext.current.getActivity()
+
+    // All Thread Network actions are handled by UI State class ThreadNetworkUiState.
+    // We are not using a ViewModel because ThreadNetwork actions require Activity
+    // and the lifetime of a ViewModel is longer than Activity.
+    val threadNetworkUiState = remember(activity) { ThreadNetworkUiState(activity!!, viewModel) }
+
+    // All calls into the ThreadNetworkUiState are encapsulated within this lambda which makes
+    // it possible to avoid exposing ThreadNetworkUiState into the composable ThreadScreen
+    // (which is less optimal for recompositions as there is a lot of stuff in ThreadNetworkUiState).
+    // ActionRequest provides all the information needed by ThreadNetworkUiState
+    // to properly process the Action triggered by the user.
+    //
+    // Note that this makes it challenging to handle UIState functions that are suspendable.
```

Comment:
For this, if you did `ThreadNetworkUiState(activity!!, rememberCoroutineScope(), viewModel)`, would the passed in coroutine scope help in launching suspending functions in the `ThreadNetworkUiState`?

Reply:
Good point. I wasn't sure if it would be fine to pass that coroutine scope to ThreadNetworkUiState. Added a TODO to test this out in a future PR.